third_party/eigen3/eigen_archive.BUILD

```diff
@@ -55,8 +55,10 @@ cc_library(
         # This define (mostly) guarantees we don't link any problematic
         # code. We use it, but we do not rely on it, as evidenced above.
         "EIGEN_MPL2_ONLY",
-        "EIGEN_MAX_ALIGN_BYTES=64",
-    ],
+    ] + select({
```

Comment:
I'm quite hesitant to do this.

It will likely affect performance for CPU ops (particular on intel), and can lead to ODR issues if other plugins/custom ops are built with a different flag.  We also currently _need_ at least 16-byte alignment for CUDA kernels - so I'm not sure if anyone is building TF for macos with nvidia GPU support.

The 64-byte alignment is meant to be compatible with aligned AVX512 instructions on intel CPUs.  For ARM, we could lower this to 16 (although again, we would need to warn all users/other plugins about potential ODR issues, and that they will need to recompile with these new flags).

Reply:
@cantonios the issue is that we have a custom memory allocator where the Plugin doesn't have any way to set the Alignment requirement and it currently raises error in `tensor.h`. We can restrict this to MacOS ARM64 builds. But we need to disable this Alignment requirement otherwise it asserts on many of our Networks with Metal plugin.

Reply:
Why can't you make your allocator allocate aligned memory blocks?  Does it work if you set the alignment to 16 (or remove the flag altogether, which is the same thing)?

The failing tests mean that the metal plugin is allocating CPU memory for a tensor, then you call an Eigen-based CPU kernel on it.  Is this intended?

Reply:
>Why can't you make your allocator allocate aligned memory blocks? Does it work if you set the alignment to 16 (or remove the flag altogether, which is the same thing)?

The failure comes from this [check](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.h#L316). The ptr alignment check fails. This is all happening for Device caching allocator. Even if we set the Memory allocator align to some memory block this will fail. I think we can set the EIGEN_MAX_ALIGN_BYTES to 0.

>The failing tests mean that the metal plugin is allocating CPU memory for a tensor, then you call an Eigen-based CPU kernel on it. Is this intended?

No, this is all happening for Device caching allocator. The Host allocations are meeting the alignment requirement.

Reply:
That alignment check should only be called when casting a TF tensor to an Eigen tensor, like [here](https://github.com/tensorflow/tensorflow/blob/7ed09da40ae91c566e220fdddc5a2c7db3b7e6ad/tensorflow/core/framework/tensor.h#L770):
```cpp
template <typename T, size_t NDIMS>
typename TTypes<T, NDIMS>::Tensor Tensor::tensor() {
  CheckTypeAndIsAligned(DataTypeToEnum<T>::v());
  return typename TTypes<T, NDIMS>::Tensor(base<T>(),
                                           shape().AsEigenDSizes<NDIMS>());
}
```
If you check the [type](https://github.com/tensorflow/tensorflow/blob/a2f8649ab468be6353ddefd84f742bcaf6606cf9/tensorflow/core/framework/tensor_types.h#L29) of that `TTypes<T,NDIMS>::Tensor`, it is an explicitly-aligned map:
```cpp
typedef Eigen::TensorMap<Eigen::Tensor<T, NDIMS, Eigen::RowMajor, IndexType>,
                         Eigen::Aligned>  // <--- EXPECTS A SPECIFIC POINTER ALIGNMENT
      Tensor;
```
If any of these casts to Eigen tensors are called, the tensor needs to be aligned.  If no casts to Eigen tensors are called, they don't need to be aligned.

What function ends up calling these alignment checks for your device tensors?  It's possible we have extraneous alignment checks that don't need to be there.  Otherwise, is there a cast to an Eigen tensor that doesn't need to be there, or do you actually end up calling some Eigen function?

> I think we can set the EIGEN_MAX_ALIGN_BYTES to 0.

That's the same as `EIGEN_DONT_ALIGN`.  I don't think you want this.

Preferably, if you have an unaligned tensor on device, you would just never cast to an Eigen tensor, and never trigger the alignment check.  If there is a cast to an Eigen tensor that's necessary, but isolated, you can instead cast to an [`UnalignedTensor`](https://github.com/tensorflow/tensorflow/blob/a2f8649ab468be6353ddefd84f742bcaf6606cf9/tensorflow/core/framework/tensor_types.h#L37).  But if you do end up calling existing Eigen-based kernels, then your GPU memory allocator should allocate aligned memory blocks.